transports/janus_websockets.c

```diff
@@ -1198,56 +1201,70 @@ static int janus_websockets_common_callback(
+					if (!response) {
+						/* No messages found */
+						janus_mutex_unlock(&ws_client->ts->mutex);
+						return 0;
+					}
+					if (g_atomic_int_get(&ws_client->destroyed) || g_atomic_int_get(&stopping)) {
+						free(response);
+						janus_mutex_unlock(&ws_client->ts->mutex);
+						return 0;
 					}
-					/* Done for this round, check the next response/notification later */
-					lws_callback_on_writable(wsi);
-					janus_mutex_unlock(&ws_client->ts->mutex);
-					return 0;
-				}
-				/* Shoot all the pending messages */
-				char *response = g_async_queue_try_pop(ws_client->messages);
-				if(response && !g_atomic_int_get(&ws_client->destroyed) && !g_atomic_int_get(&stopping)) {
 					/* Gotcha! */
-					int buflen = LWS_PRE + strlen(response);
+					JANUS_LOG(LOG_HUGE, "[%s-%p] Sending WebSocket message (%zu bytes)...\n", log_prefix, wsi, strlen(response));
+					size_t buflen = LWS_PRE + strlen(response);
 					if (buflen > ws_client->buflen) {
 						/* We need a larger shared buffer */
-						JANUS_LOG(LOG_HUGE, "[%s-%p] Re-allocating to %d bytes (was %d, response is %zu bytes)\n", log_prefix, wsi, buflen, ws_client->buflen, strlen(response));
+						JANUS_LOG(LOG_HUGE, "[%s-%p] Re-allocating to %zu bytes (was %zu, response is %zu bytes)\n", log_prefix, wsi, buflen, ws_client->buflen, strlen(response));
 						ws_client->buflen = buflen;
 						ws_client->buffer = g_realloc(ws_client->buffer, buflen);
 					}
 					memcpy(ws_client->buffer + LWS_PRE, response, strlen(response));
-					JANUS_LOG(LOG_HUGE, "[%s-%p] Sending WebSocket message (%zu bytes)...\n", log_prefix, wsi, strlen(response));
-					int sent = lws_write(wsi, ws_client->buffer + LWS_PRE, strlen(response), LWS_WRITE_TEXT);
-					JANUS_LOG(LOG_HUGE, "[%s-%p]   -- Sent %d/%zu bytes\n", log_prefix, wsi, sent, strlen(response));
-					if(sent > -1 && sent < (int)strlen(response)) {
-						/* We couldn't send everything in a single write, we'll complete this in the next round */
-						ws_client->bufpending = strlen(response) - sent;
-						ws_client->bufoffset = LWS_PRE + sent;
-						JANUS_LOG(LOG_HUGE, "[%s-%p]   -- Couldn't write all bytes (%d missing), setting offset %d\n",
-							log_prefix, wsi, ws_client->bufpending, ws_client->bufoffset);
-					}
+					/* Initialize pending bytes count and buffer offset */
+					ws_client->bufpending = strlen(response);
+					ws_client->bufoffset = LWS_PRE;
 					/* We can get rid of the message */
 					free(response);
-					/* Done for this round, check the next response/notification later */
-					lws_callback_on_writable(wsi);
+				}
+
+				if (g_atomic_int_get(&ws_client->destroyed) || g_atomic_int_get(&stopping)) {
 					janus_mutex_unlock(&ws_client->ts->mutex);
 					return 0;
 				}
+
+				/* Evaluate amount of data to send according to MESSAGE_CHUNK_SIZE */
+				int amount = ws_client->bufpending <= MESSAGE_CHUNK_SIZE ? ws_client->bufpending : MESSAGE_CHUNK_SIZE;
```

Comment:
nit: why not use `size_t` for amount (to avoid casting later) as it seems it's never needed as an int.

Reply:
Also a more descriptive name like: `amount_to_send` or `bytes_to_send` might be preferable.

Reply:
Honestly I've used the same name (`amount`) and type (`int`) present in the [lws example](https://github.com/warmcat/libwebsockets/blob/ec9e8ab892f2dc569ae09336c80c2796fd14a6ff/minimal-examples/ws-server/minimal-ws-server-pmd-bulk/protocol_lws_minimal_pmd_bulk.c#L91).

Reply:
@atoppi sure but in their case they don't need to do the casts. It's a minor point but cleaner IMO.

Reply:
If `amount` type is set to `size_t` the following simple test will fail
```c
#include <stdio.h>
#include <inttypes.h>
#include <stdlib.h>
#include <limits.h>

int main(int argc, char *argv[]) {	
    size_t amount = 2400;
    /* Simulate error in lws_write */
    int sent = -1;
    if (sent < amount) {
        printf("sent < amount !\n");
        return 0;
    }
    return -1;
}
```

Any cast in the comparison ` if (sent < amount)` might have a problem:
- if we do `(size_t)sent`, we are casting a signed to unsigned
- if we do `(int)amount`, we are casting to a lower (at least on my machine) size type (8 vs 4)

Am I missing something ?

Reply:
So in that example, since you are comparing to signed you should be using `ssize_t` for amount, but AFAICT in your patch you were only ever comparing to unsigned.

Reply:
[here](https://github.com/meetecho/janus-gateway/blob/ws-send-fragments/transports/janus_websockets.c#L1249
) I am comparing `sent` (type `int`, return value of `lws_write`) with `amount`.
According to lws docs about `lws_write`:
```
Return may be -1 for a fatal error needing connection close
```

Reply:
Right so `ssize_t` would make sense then.

Reply:
I personally don't see the point: `lws_write` returns `int`, not `ssize_t`, and comparing `size_t` and `ssize_t` would result in a comparison between values with different sizes anyway. Using `int` for that part seems reasonable.

Reply:
Yeah didn't want to derail the discussion here as we know these values are capped pretty low anyway.